modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +330,168 @@ void run(ClusterState state) {
+                            // at this point the source index is part of the data stream and the downsample index is complete but not
+                            // part of the data stream. we need to replace the source index with the downsample index in the data stream
+                            // and delete the source index.
+                            if (dataStream.getIndices().contains(downsampleIndexMeta.getIndex()) == false) {
+                                affectedIndices.add(backingIndex.getIndex());
+                                clusterStateChangesDeduplicator.executeOnce(
+                                    new ReplaceSourceWithDownsampleIndexTask(
+                                        dataStream.getName(),
+                                        index.getName(),
+                                        downsampleIndexName,
+                                        null
+                                    ),
+                                    new ErrorRecordingActionListener(indexName, errorStore),
+                                    (req, reqListener) -> replaceSourceWithDownsampleIndexAndDeleteSource(
+                                        new ReplaceSourceWithDownsampleIndexTask(
+                                            dataStream.getName(),
+                                            index.getName(),
+                                            downsampleIndexName,
+                                            reqListener
+                                        ),
+                                        reqListener
+                                    )
+                                );
+                                break;
+                            }
+                        }
+                    } else if (round.equals(lastRound)) {
+                        // no maintenance needed for previously started downsampling actions, so kick off the last matching round
+                        affectedIndices.add(index);
+                        DownsampleAction.Request request = new DownsampleAction.Request(
+                            indexName,
+                            downsampleIndexName,
+                            null,
+                            round.config()
+                        );
+                        transportActionsDeduplicator.executeOnce(
+                            request,
+                            new ErrorRecordingActionListener(indexName, errorStore),
+                            (req, reqListener) -> downsampleIndex(request, reqListener)
+                        );
+                    }
+                }
+            }
         }
+        return affectedIndices;
+    }
+
+    private static List<Index> getTargetIndices(DataStream dataStream, Set<Index> indicesToExcludeForRemainingRun) {
+        return dataStream.getIndices().stream().filter(index -> indicesToExcludeForRemainingRun.contains(index) == false).toList();
     }
```

Comment:
Why don't we move the `dataStream.isIndexManagedByDataStreamLifecycle(index, state.metadata()::index)` check into this method, so that we don't have to duplicate that check inside each of the `maybeExecute…` methods?

Also, can you add a javadoc to this method?

Reply:
++ great shout